Use meshlab to fix model on Linux/Mac

buzzhuzz opened the pull request:
# Description

Fixes #5013, #4923

This change introduces new back end for the "Fix model" feature: MeshLabServer.

On Linux and Mac platforms OrcaSlicer checks if "meshlabserver" could be found in path and allows broken objects to be fixed if it is.

Simple script for meshlab added to resources.

In order to make this feature work you need to have MeshLab installed and "meshlabserver" available in PATH.

# Screenshots/Recordings/Graphs


https://github.com/user-attachments/assets/25aa1bb8-2a7c-404f-814a-c5b971b8e4e0


## Tests

Tested on Linux host only, with native build (no appimage).

buzzhuzz committed:
fix model: rename to generic names

buzzhuzz committed:
Preparations for meshlab introduction

buzzhuzz committed:
meshlab boilerplate code

buzzhuzz committed:
Proper meshlabserver presence detection

buzzhuzz committed:
Fix model using meshlabserver

buzzhuzz committed:
Meaningful message on repair model availability

buzzhuzz committed:
meshlab: Capture stderr output

buzzhuzz committed:
Fix windows build

buzzhuzz commented:
It is expected that someone proficient in MeshLab and meshes manipulation will update the meshlab script within resources directory.

I also have not tested this under Mac. From old times I remember apps are distributed in bundles and it might be tricky to add bundled app binary into the path, so I expect someone on Mac would test this change so we could update documentation/notes on how to make it working under OSX.

buzzhuzz committed:
Fix windows build

buzzhuzz committed:
Message cleanup

buzzhuzz committed:
Include lost header

sschwetz commented:
I have installed Meshlab 2023.12, and it no longer has meshlabserver. This has been replaced by PyMeshlab.

buzzhuzz commented:
Well, it looks like PyMeshLab is the way.

Originally I was going to use PyMeshLab, but switched to meshlabserver due to much easier implementation. Somhow I've missed that meshlabserver got removed in latest version.

I'm going to convert this PR to draft just by now in case someone would like to try this on linux.

I'll rework this to use PyMeshLab and update PR.

buzzhuzz committed:
meshlab: rework to use PyMeshLab

FixModel reimplemented to use PyMeshLab instead of MeshLabServer

NOTE: this functionality is still in development phase, so
      some additional configuration outside of OrcaSlicer is needed

Current change have been tasted on Linux using PyMeshLab==2024.12
Sample meshlab script created using MeshLab 2024.12.

Testing:

    1. Prepare virtual environment with PyMeshLab installed.
       This step may be skipped if pymeshlab installed globally

        $ python -m venv meshlab
        $ ./meshlab/bin/activate
        (pymeshlab) $ pip install pymeshlab

    2. (optionally) Activate virtual env

        $ source path/to/meshlab-venv/bin/activate

    3. Run OrcaSlicer from command line with venv activated:

        (pymeshlab) $ ./path/to/OrcaSlicer.AppImage

Test sequence may need to be adjusted for the Mac platform

buzzhuzz commented:
@sschwetz , would you mind testing updated changes with PyMeshLab support?

Please refer to the change description regarding testing sequence. This is not final change as we need to came to solution how to simplify configuration to skip venv usage.

buzzhuzz committed:
Generalize download path config code

buzzhuzz committed:
pymeshlab path config entry

buzzhuzz committed:
Use meshlab path config for repair

buzzhuzz committed:
Merge remote-tracking branch 'upstream/main' into dbuzz/meshlab

buzzhuzz commented:
I've added config option to set PyMeshLab location (which is recommended to be installed inside venv).
Now user may create virtualenv with PyMeshLab installed in it. OrcaSlicer shall be configured to point into /bin/ directory (the one containing activate.sh script and/or python3 link) of the respective virtualenv:

![Screenshot from 2025-01-27 01-42-21](https://github.com/user-attachments/assets/73008c3c-3fbf-493a-9ff9-8c20af5ba131)

oskardotglobal commented:
Currently trying this under macOS 15.2. 

I've grabbed the latest OrcaSlicer from the build artifacts, created a virtualenv (`python3 -m venv ~/.venv`) and built pymeshlab from source (`pip3 install git+https://github.com/cnr-isti-vclab/PyMeshLab`) since `pip3 install pymeshlab` didn't work. 

I set the path in the settings but it still doesn't seem to work (I get an error saying PyMeshLab not found). Can I provide any logs to help?

buzzhuzz commented:
@oskardotglobal , WOW I really happy it is working on OSX. Thank you for your testing.

buzzhuzz committed:
fix model documentation

oskardotglobal commented on doc/Fix-Model.md:
```diff
@@ -0,0 +1,30 @@
+    # create virtualenv directory
+    python3 -m venv pymeshlab
+    # activate virtualenv
+    source pymeshlab/bin/activate.sh
```
This has to be `bin/activate`, not `bin/activate.sh`

buzzhuzz replied:
Thanks, I've pushed the update.

buzzhuzz committed:
Fix documentation

oskardotglobal commented on doc/Fix-Model.md:
````diff
@@ -0,0 +1,30 @@
+    \
+    Simple installation may be done in next steps:
+    ```    
+    # create virtualenv directory
+    python3 -m venv pymeshlab
+    # activate virtualenv
+    source pymeshlab/bin/activate
+    # install pymeshlab
+    pip3 install pymeshlab
````
I don't think `pipx` will work unless the `pymeshlab` packages exports an executable (which I don't think it does)

buzzhuzz replied:
It might be set up this way, however using pipx system-wide may cause conflicts with distro version of python modules.

danpf commented:
The pipeline works on M1 mac OSX 14.5

but doesn't seem to be able to repair this stl - Not sure if there needs to be knobs to tweak for every object or if it should be universal. 
<img width="351" alt="image" src="https://github.com/user-attachments/assets/977456e9-9ca0-4214-89fa-bb46d7960360" />

[czfa_flexframe_screw.stl.zip](https://github.com/user-attachments/files/18815803/czfa_flexframe_screw.stl.zip)

from https://www.printables.com/model/647794-flexible-sturdy-phone-arm-100-printed

klylabs commented:
Can't we just setup the env for the user automatically?

Since ubuntu/debian is the officially supported distro, and those two comes with python preinstalled, wouldn't it be safe to assume that python will be available in whatever env orca-slicer is run in?

SoftFever committed:
Merge branch 'main' into dbuzz/meshlab

SoftFever commented:
I have mixed feelings about this approach. On one hand, it really fills the gap in terms of functionality. But on the other hand, this solution involves too much of an environment that is out of control. Also, it feels not integrated. So I think I will not incorporate this into the main branch.
But I commend PR.
Thank you

EliteTK commented:
@buzzhuzz great work!

@SoftFever this patch is effectively rather indirectly via python -> pymeshlab -> meshlab calling VCGlib to do the actual mesh fixing.

Would you accept a patch to just (optionally) depend on VCGlib instead? I am probably going to write a patch for it anyway for my own purposes but an answer now would help me gauge how much effort I should put into polishing it.

yw4z commented:
@ianalexis i assume this can be closed with merge of [#12155](https://github.com/OrcaSlicer/OrcaSlicer/pull/12155)
